src/inmanta/ast/statements/__init__.py

```diff
@@ -149,54 +193,258 @@ def get_dataflow_node(self, graph: DataflowGraph) -> dataflow.NodeReference:
 
 
 class Resumer(ExpressionStatement):
+    """
+    Resume on a set of requirement variables' values when they become ready (i.e. they are complete).
+    """
+
     def resume(self, requires: Dict[object, object], resolver: Resolver, queue: QueueScheduler, target: ResultVariable) -> None:
         pass
 
 
 class RawResumer(ExpressionStatement):
+    """
+    Resume on a set of requirement variables when they become ready (i.e. they are complete).
+    """
+
     def resume(self, requires: Dict[object, ResultVariable], resolver: Resolver, queue: QueueScheduler) -> None:
         pass
 
 
+class VariableReferenceHook(RawResumer):
+    """
+    Generic helper class for adding a hook to a variable (ResultVariable) object. Supports both plain variables and instance
+    attributes. Calls variable resumer with the variable object as soon as it's available.
+    This class is not a full AST node, rather it is a Resumer only. It is meant to delegate common resumer behavior that would
+    otherwise need to be implemented as custom resumer logic in each class that needs it.
+    """
+
+    def __init__(
+        self,
+        instance: Optional["Reference"],
+        name: str,
+        variable_resumer: "VariableResumer",
+    ) -> None:
+        super().__init__()
+        self.instance: Optional["Reference"] = instance
+        self.name: str = name
+        self.variable_resumer: "VariableResumer" = variable_resumer
+
+    def schedule(self, resolver: Resolver, queue: QueueScheduler) -> Waiter:
+        """
+        Schedules this instance for execution. Waits for the variable's requirements before resuming.
+        """
+        return RawUnit(
+            queue,
+            resolver,
+            # no need for gradual execution here because this class represents an attribute reference on self.instance,
+            # which is not allowed on multi variables (the only kind of variables that would benefit from gradual execution)
+            self.instance.requires_emit(resolver, queue) if self.instance is not None else {},
```

Comment:
If I understand correctly, if the instance is None, the resume method can be called directly here? (there are no requires)?

Reply:
Indeed. My assumption was that the impact of creating the unit anyway would be very limited and I found it simplified the logic (better stick to the semantics of the resume method). But I have no strong attachment to this approach so if you prefer I can call it directly.